examples/generative/dreambooth.py

```diff
@@ -0,0 +1,672 @@
+            )
+            loss = self.compute_loss(target, model_pred)
+            if self.use_mixed_precision:
+                loss = self.optimizer.get_scaled_loss(loss)
+
+        # Update parameters of the diffusion model.
+        trainable_vars = self.diffusion_model.trainable_variables
+        gradients = tape.gradient(loss, trainable_vars)
+        if self.use_mixed_precision:
+            gradients = self.optimizer.get_unscaled_gradients(gradients)
+        gradients = [tf.clip_by_norm(g, self.max_grad_norm) for g in gradients]
+        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
+
+        return {m.name: m.result() for m in self.metrics}
+
+    def get_timestep_embedding(self, timestep, dim=320, max_period=10000):
+        half = dim // 2
+        log_max_preiod = tf.math.log(tf.cast(max_period, tf.float32))
+        freqs = tf.math.exp(
+            -log_max_preiod * tf.range(0, half, dtype=tf.float32) / half
+        )
+        args = tf.convert_to_tensor([timestep], dtype=tf.float32) * freqs
+        embedding = tf.concat([tf.math.cos(args), tf.math.sin(args)], 0)
+        return embedding
+
+    def sample_from_encoder_outputs(self, outputs):
+        mean, logvar = tf.split(outputs, 2, axis=-1)
+        logvar = tf.clip_by_value(logvar, -30.0, 20.0)
+        std = tf.exp(0.5 * logvar)
+        sample = tf.random.normal(tf.shape(mean), dtype=mean.dtype)
+        return mean + std * sample
+
+    def compute_loss(self, target, model_pred):
+        # Chunk the noise and model_pred into two parts and compute the loss
+        # on each part separately.
+        # Since the first half of the inputs has instance samples and the second half
+        # has class samples, we do the chunking accordingly.
+        model_pred, model_pred_prior = tf.split(
+            model_pred, num_or_size_splits=2, axis=0
+        )
+        target, target_prior = tf.split(target, num_or_size_splits=2, axis=0)
+
+        # Compute instance loss.
+        loss = self.compiled_loss(target, model_pred)
+
+        # Compute prior loss.
+        prior_loss = self.compiled_loss(target_prior, model_pred_prior)
+
+        # Add the prior loss to the instance loss.
+        loss = loss + self.prior_loss_weight * prior_loss
+        return loss
+
+    def save_weights(self, filepath, overwrite=True, save_format=None, options=None):
+        # Overriding this method will allow us to use the `ModelCheckpoint`
+        # callback directly with this trainer class. In this case, it will
+        # only checkpoint the `diffusion_model` since that's what we're training
+        # during fine-tuning.
+        self.diffusion_model.save_weights(
+            filepath=filepath,
+            overwrite=overwrite,
+            save_format=save_format,
+            options=options,
+        )
+
+
+"""
+## Trainer initialization
+"""
+
+# Comment it if you are not using a GPU having tensor cores.
+tf.keras.mixed_precision.set_global_policy("mixed_float16")
```

Comment:
You might want to put this line much earlier for safety (fine in this case though)

Reply:
Yes, from our codebase, we [do it](https://github.com/sayakpaul/dreambooth-keras/blob/895f3c07665e0a6e0e392c8a9fabadd13c58674a/train_dreambooth.py#L132) just after the TensorFlow import.